lib/ably/realtime/channel.rb

```diff
@@ -179,15 +179,28 @@ def presence
         @presence
       end
 
-      # Return the message history of the channel
+      # Return the message history of the channel.
+      #
+      # Once attached to a channel, you can retrieve messages published on the channel before the
+      # channel was attached with the option <tt>:end => :before_attach</tt>.  This is very useful for
+      # developers who wish to display both new real-time messages as well as historical messages with
+      # the guarantee that no messages have been missed.
+      #
       #
       # @param (see Ably::Rest::Channel#history)
       # @option options (see Ably::Rest::Channel#history)
+      # @option options [Integer,Time,Symbol]   :end        Time or millisecond since epoch, or +:before_attach+ for all messages published before this channel has been attached.
       #
       # @yield [Ably::Models::PaginatedResource<Ably::Models::Message>] An Array of {Ably::Models::Message} objects that supports paging (#next_page, #first_page)
       #
       # @return [Ably::Util::SafeDeferrable]
       def history(options = {}, &callback)
+        if options[:end] == :before_attach
+          raise ArgumentError, 'option :before_attach cannot be specified if the channel is not attached' unless attached?
```

Comment:
> it seems a bit weird that someone is asking for history prior to the channel attaching when the channel is not even attached

The whole point is that you can subscribe, which means that messages from the moment of subscription are sent to your listener, and then this functionality allows you to get exactly the messages that occurred before you attached. So it is only meaningful if there is a call to `subscribe()` prior to the call to `history()` using this param. Since the `subscribe()` implicitly calls `attach()` there is no sense in the `attach()` also being called later.

> We should just log it and warn them that they are using this option incorrectly. Not potentially break their whole app.

An exception isn't an inconvenience to a programmer; it's a way to tell him that the program he wrote cannot meaningfully continue. So if an exception is called for, it helps him.